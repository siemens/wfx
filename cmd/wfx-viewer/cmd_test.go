package main

import (
	"bytes"
	"os"
	"testing"

	approvals "github.com/approvals/go-approval-tests"
	"github.com/siemens/wfx/workflow/dau"
	"github.com/stretchr/testify/require"
	"gopkg.in/yaml.v3"
)

func TestPlantUML_Stdout(t *testing.T) {
	f := rootCmd.PersistentFlags()
	_ = f.Set(outputFlag, "")
	_ = f.Set(outputFormatFlag, "plantuml")

	buf := new(bytes.Buffer)
	rootCmd.SetOut(buf)

	tmpFile, _ := os.CreateTemp(os.TempDir(), "TestPlantUML.*")
	b, _ := yaml.Marshal(dau.DirectWorkflow())
	_, _ = tmpFile.Write(b)
	tmpFile.Close()
	defer os.Remove(tmpFile.Name())

	rootCmd.SetArgs([]string{tmpFile.Name()})
	err := rootCmd.Execute()
	require.NoError(t, err)

	approvals.VerifyString(t, buf.String())
}

func TestPlantUML_File(t *testing.T) {
	f := rootCmd.PersistentFlags()

	outFile, _ := os.CreateTemp(os.TempDir(), "TestPlantUML.*")
	defer os.Remove(outFile.Name())

	_ = f.Set(outputFlag, outFile.Name())
	_ = f.Set(outputFormatFlag, "plantuml")

	tmpFile, _ := os.CreateTemp(os.TempDir(), "TestPlantUML.*")
	b, _ := yaml.Marshal(dau.DirectWorkflow())
	_, _ = tmpFile.Write(b)
	tmpFile.Close()
	defer os.Remove(tmpFile.Name())

	err := rootCmd.RunE(rootCmd, []string{tmpFile.Name()})
	require.NoError(t, err)

	content, _ := os.ReadFile(outFile.Name())
	approvals.VerifyString(t, string(content))
}
