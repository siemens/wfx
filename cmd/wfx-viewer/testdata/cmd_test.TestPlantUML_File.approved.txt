@startuml
state INSTALL as "<color:black>INSTALL</color>" #00cc00: instruct client to start installation
state INSTALLING as "<color:black>INSTALLING</color>" #00cc00: installation progress update from client
state INSTALLED as "<color:black>INSTALLED</color>" #00cc00: client signaled installation success
state ACTIVATE as "<color:black>ACTIVATE</color>" #00cc00: instruct client to start activation
state ACTIVATING as "<color:black>ACTIVATING</color>" #00cc00: client activates update
state ACTIVATED as "<color:black>ACTIVATED</color>" #4993dd: client signaled activation success
state TERMINATED as "<color:black>TERMINATED</color>" #9393dd: client aborted update with error
INSTALL --> INSTALLING: CLIENT
INSTALL --> TERMINATED: CLIENT
INSTALLING --> INSTALLING: CLIENT
INSTALLING --> TERMINATED: CLIENT
INSTALLING --> INSTALLED: CLIENT
INSTALLED --> ACTIVATE: WFX [IMMEDIATE]
ACTIVATE --> ACTIVATING: CLIENT
ACTIVATE --> TERMINATED: CLIENT
ACTIVATING --> ACTIVATING: CLIENT
ACTIVATING --> TERMINATED: CLIENT
ACTIVATING --> ACTIVATED: CLIENT
legend right
  | Color | Group | Description |
  | <#00cc00> | OPEN | regular workflow-advancing states |
  | <#4993dd> | CLOSED | a successful update's terminal states |
  | <#9393dd> | FAILED | a failed update's terminal states |
  | <#000000> |  | The state doesn't belong to any group. |
endlegend
@enduml
